Add an action method for creating a distance request

tgolen opened the pull request:
### Details
This PR adds a new method to the IOU action file for creating a distance request. Since the majority of the logic was shared with `requestMoney()`, much of the code needed to be DRYed and moved to another method.

### Fixed Issues
$ https://github.com/Expensify/App/issues/22711


### Tests
**Note: This logic is not connected to anything at the moment so the only test to do is to ensure that there are no regressions when requesting money.**

1. Click the + button
2. Select "Request Money"
3. Go through the flow and verify that it works without any regressions

- [ ] Verify that no errors appear in the JS console

### Offline tests
None

### QA Steps
Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->
![image](https://github.com/Expensify/App/assets/1228807/3696f21e-bef2-4826-9eac-632209cef4c0)


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->
![image](https://github.com/Expensify/App/assets/1228807/4163b315-2feb-456c-9389-c727d42c5f34)

</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->
![image](https://github.com/Expensify/App/assets/1228807/f8ec7a13-0e3c-4fcc-9c8f-c2d4d48a92e3)


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->
![image](https://github.com/Expensify/App/assets/1228807/f174a293-9fcf-4f8c-854b-f49eb85a4ddb)

</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->
![image](https://github.com/Expensify/App/assets/1228807/b7d47565-8cd0-4a9d-ac56-7e5e115e9c18)


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->
![image](https://github.com/Expensify/App/assets/1228807/5064e81d-6df3-47ca-accd-77852a4f51ac)

</details>

tgolen committed:
Add initial method and DRY up logic for IOU report

tgolen committed:
Move more logic to DRY method

tgolen committed:
Move more logic into DRY method

tgolen committed:
Add JS Docs and remove unused variables

tgolen committed:
Simplify params, add more JS Docs

tgolen committed:
Merge branch 'main' into tgolen-createdistancerequest

tgolen commented:
@neil-marcellini Would you mind pulling this branch and seeing if there are any regressions when requesting money? I tried this over and over again in my environment and I ran into so many server errors that I couldn't get this tested. I'm thinking I need to rebuild my VM completely to try and get it tested.

tgolen commented:
Nevermind, it looks like rebuilding my VM and starting with a fresh database works 👍

tgolen committed:
Fix the onyx data

tgolen committed:
Merge branch 'main' into tgolen-createdistancerequest

tgolen commented:
This is ready for review, but I am having a difficult time getting anything to run on Android, so that is where my testing is being held-up. @robertKozik maybe you are able to have better luck testing this on Android than me!

tgolen commented:
I finally got Android running and finished testing this.

tgolen committed:
Merge branch 'main' into tgolen-createdistancerequest

neil-marcellini commented on src/libs/actions/IOU.js:
```diff
@@ -364,7 +377,7 @@ function requestMoney(report, amount, currency, payeeEmail, payeeAccountID, part
         receiptObject.state = CONST.IOU.RECEIPT_STATE.SCANREADY;
     }
-    const optimisticTransaction = TransactionUtils.buildOptimisticTransaction(amount, currency, iouReport.reportID, comment, '', '', undefined, receiptObject);
+    const transaction = TransactionUtils.buildOptimisticTransaction(amount, currency, iouReport.reportID, comment, '', '', undefined, receiptObject);
```
For distance requests we create the transaction optimistically as soon as the flow starts, so we need to get that transaction and merge it with the other optimistic transaction data built here.
https://github.com/Expensify/App/blob/82633312c96994efaf664fbea4ed2bff598c986e/src/pages/iou/DistanceRequestPage.js#L29

tgolen replied:
OK, it looks like `createEmptyTransaction()` is a method that generates a random transactionID (I think maybe it is a poorly named method). In order to use that transactionID, I have modified `createDistanceRequest()` to accept an existing transaction ID. This ID is then used when creating the optimistic transaction in this code instead of generating a new random ID.

I think that should be all that is needed for when someone hooks up to the `createDistanceRequest()` method, but please double-check my assumptions.

neil-marcellini replied:
We can build the optimistic transaction with the same ID, but then we still need to merge it with the transaction that already exists in Onyx. The existing transaction will have waypoints set on it, and maybe some other data that I don't think we want to lose.

neil-marcellini reviewed (changes requested):
Looking good and tests well. I realized we need to make one modification regarding the optimistic transaction.

tgolen committed:
Merge branch 'main' into tgolen-createdistancerequest

tgolen committed:
Use existing transaction ID when generating optimistic transaction

neil-marcellini reviewed (changes requested):
Looking better, but as I mentioned in an earlier thread I think we should merge the optimistic transaction with the one that is currently stored in Onyx.

tgolen committed:
Merge the transaction data instead of setting it

tgolen committed:
Merge branch 'main' into tgolen-createdistancerequest

tgolen committed:
Always merge, don't set

neil-marcellini reviewed (dismissed):
Looks good given that the optimistic transaction is merged with the one that currently exists.

tgolen committed:
Fix typo in comment

tgolen committed:
Remove comment

tgolen commented on src/libs/actions/IOU.js:
```diff
@@ -101,7 +101,7 @@ function buildOnyxDataForMoneyRequest(
         {
             // Use SET for new reports because it doesn't exist yet, is faster and we need the data to be available when we navigate to the chat page
-            onyxMethod: isNewChatReport ? Onyx.METHOD.SET : Onyx.METHOD.MERGE,
+            onyxMethod: Onyx.METHOD.MERGE,
```
These are being changed because:
1. Set and merge shouldn't be mixed. It will lead to undefined behavior of the order in which subscribers are updated (ie. you end up seeing a little flash of incorrect data as amounts go from $0 to whatever was set in the request).
2. Using `set` just because "it's faster" is not a reason to use set

I checked with @luacmartins who originally made the change and we agreed it was OK to use merge for everything for now.

allroundexperts reviewed (approved):
Tests well!

tgolen commented:
> In offline mode, requesting money from a new user results in the following console error:

I was not able to reproduce this one.


https://github.com/Expensify/App/assets/1228807/378ecc1e-6fed-474c-9113-2ade5178f6b0

tgolen committed:
Change more calls to merge

tgolen committed:
Revert "Change more calls to merge"

This reverts commit e83cb43a8879207a9d78328411d0b579682780ca.

tgolen committed:
Update IOU.js

tgolen committed:
Manually merge transaction data from existing transaction

tgolen committed:
Fix variable references

tgolen commented:
OK, I've reverted a lot of the `set/merge` changes because that was too much of a rabbit hole for now. I've gone ahead and added code which manually merges the existingTransaction (if there is one) with the optimisticTransaction to ensure the final transaction object has all the correct data. The two tests I did to ensure this was working are:

1. Go offline
2. Create a money request
3. Verify there is no flash of "not found" page or the "skeleton UI" loading page
4. Go online
5. Create a money request
3. Verify there is no flash of "not found" page or the "skeleton UI" loading page

allroundexperts reviewed (approved):
Looking good. Tested again both in offline and online mode. Did not see anything broken!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5931749076) to staging by https://github.com/neil-marcellini in version: 1.3.56-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5958899174) to production by https://github.com/roryabraham in version: 1.3.56-24 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅
